lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1640 @@
+//! The top-level network map tracking logic lives here.
+
+use secp256k1::key::PublicKey;
+use secp256k1::Secp256k1;
+use secp256k1;
+
+use bitcoin_hashes::sha256d::Hash as Sha256dHash;
+use bitcoin_hashes::Hash;
+use bitcoin::blockdata::script::Builder;
+use bitcoin::blockdata::opcodes;
+
+use chain::chaininterface::{ChainError, ChainWatchInterface};
+use ln::features::{ChannelFeatures, NodeFeatures};
+use ln::msgs::{DecodeError,ErrorAction,LightningError,RoutingMessageHandler,NetAddress};
+use ln::msgs;
+use util::ser::{Writeable, Readable, Writer, ReadableArgs};
+use util::logger::Logger;
+
+use std::cmp;
+use std::sync::{RwLock,Arc};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::collections::BTreeMap;
+use std::collections::btree_map::Entry as BtreeEntry;
+use std;
+
+
+/// Receives network updates from peers to track view of the network.
+pub struct NetGraphMsgHandler {
+	secp_ctx: Secp256k1<secp256k1::VerifyOnly>,
+	/// Representation of the payment channel network
+	pub network_graph: RwLock<NetworkGraph>,
+	chain_monitor: Arc<ChainWatchInterface>,
+	full_syncs_requested: AtomicUsize,
+	logger: Arc<Logger>,
+}
+
+impl NetGraphMsgHandler {
+	/// Creates a new tracker of the actual state of the network of channels and nodes.
+	pub fn new(our_pubkey: PublicKey, chain_monitor: Arc<ChainWatchInterface>, logger: Arc<Logger>) -> Self {
+		let mut nodes = BTreeMap::new();
+		nodes.insert(our_pubkey.clone(), NodeInfo {
+			channels: Vec::new(),
+			lowest_inbound_channel_fees: None,
+			features: NodeFeatures::empty(),
+			last_update: None,
+			rgb: [0; 3],
+			alias: [0; 32],
+			addresses: Vec::new(),
+			announcement_message: None,
+		});
+		NetGraphMsgHandler {
+			secp_ctx: Secp256k1::verification_only(),
+			network_graph: RwLock::new(NetworkGraph {
+				channels: BTreeMap::new(),
+				nodes: nodes,
+			}),
+			full_syncs_requested: AtomicUsize::new(0),
+			chain_monitor,
+			logger: logger.clone(),
+		}
+	}
+
+	/// Get network addresses by node id
+	pub fn get_addresses(&self, pubkey: &PublicKey) -> Option<Vec<NetAddress>> {
+		let network = self.network_graph.read().unwrap();
+		network.get_nodes().get(pubkey).map(|n| n.addresses.clone())
+	}
+
+	fn remove_channel_in_nodes(nodes: &mut BTreeMap<PublicKey, NodeInfo>, chan: &ChannelInfo, short_channel_id: u64) {
+		macro_rules! remove_from_node {
+			($node_id: expr) => {
+				if let BtreeEntry::Occupied(mut entry) = nodes.entry($node_id) {
+					entry.get_mut().channels.retain(|chan_id| {
+						short_channel_id != *NetworkGraph::get_short_id(chan_id)
+					});
+					if entry.get().channels.is_empty() {
+						entry.remove_entry();
+					}
+				} else {
+					panic!("Had channel that pointed to unknown node (ie inconsistent network map)!");
+				}
+			}
+		}
+		remove_from_node!(chan.one_to_two.src_node_id);
+		remove_from_node!(chan.two_to_one.src_node_id);
+	}
+
+	/// Marks a node as having failed a route. This will avoid re-using the node in routes for now,
```

Comment:
Honestly, we should probably just wholesale remove this - its not implemented and the API doesn't even make sense...we need a much more wholesale routing state machine.